### ISSUE TITLE
Open source code

### ISSUE THREAD

**Issue by skeletonNN** [41xu/Your3dEmoji#1]
Hi, your work is very interesting. Please open your source code. Thank you~

**Comment by 41xu** [comment]
Thanks for your interest. I'm cleaning the code and plan to release it in the next two days(hope so...) 
If you have some problems with re-implement, just open some issues and contact me directly!
Sorry for my delay...

**Comment by skeletonNN** [comment]
Thank you. I want to know the implementation details of the formula 3.

**Comment by skeletonNN** [comment]
Hi I re-implement the formula 3. Can you give some suggestions?
`import torch
import torch.nn as nn

import lpips

import numpy as np

from models.arcface import resnet_face18

from torchvision import transforms

def cosin_metric(x1, x2):
    return np.dot(x1, x2) / (np.linalg.norm(x1) * np.linalg.norm(x2))
    
def get_featurs(model, image, label, batch_size = 8, input_shape=(1, 128, 128)):
    
    '''
        image_list: tensor b*c*h*w
        label_list: tensor b*c*h*w
    '''
    model.eval()
    
    images = None
    labels = None
    
    transform = transforms.Compose([
        transforms.ToPILImage(),
        transforms.RandomCrop(input_shape[1:]),
        transforms.Grayscale(),
        transforms.ToTensor(),
        transforms.Normalize(mean=[0.5], std=[0.5]),
    ])
    
    for img in image:
        img = transform(img)
        img = torch.unsqueeze(img, 0)
        if images == None:
            images = img
        else:
            images = torch.cat((images, img), 0)
            
    for img in label:
        img = transform(img)
        img = torch.unsqueeze(img, 0)
        if labels == None:
            labels = img
        else:
            labels = torch.cat((labels, img), 0)
            
    output_image = model(images.cuda())
    fea_image = output_image.data.cpu().numpy()
    output_label = model(labels.cuda())
    fea_label = output_label.data.cpu().numpy()
    
    cos = 0
    for i in range(batch_size):
        cos += cosin_metric(fea_image[i], fea_label[i])
    
    cos/=batch_size

    return 1 - cos


def lpips_loss(image, label, batch_size = 8):
    loss_fn_vgg = lpips.LPIPS(net="alex").cuda() # net choose # alexnet need to hub checkpoints
    loss = 0
    for i in range(batch_size):
        loss += loss_fn_vgg(image[i], label[i]).item()
        
    return loss / batch_size
    

def lid_loss(image, label):
    # https://github.com/ronghuaiyang/arcface-pytorch/blob/master/test.py
    
    config = {
        "use_se": False,
        "test_model_path": "pkl/resnet18_110.pth",
        
    }
    model = resnet_face18(config["use_se"])
    model.load_state_dict({k.replace('module.',''):v for k,v in torch.load(config["test_model_path"]).items()})
    model.cuda()
    loss = get_featurs(model, image, label)
    return loss

def l1_loss(image, label):
    loss = nn.L1Loss()
    return loss(image, label)

def imitator_loss(image, label, lambda1 = 1.0, lambda2 = 0.8, lambda3 = 1.0):
    transform_l1 = transforms.Compose([
        transforms.Normalize(mean=[0.5], std=[0.5]),
    ])
    image_l1 = transform_l1(image / 255)
    label_l1= transform_l1(label / 255)
    l1 = l1_loss(image_l1, label_l1)
    
    image_uint8 = image.to(torch.uint8)
    
    lpips = lpips_loss(image_uint8, label)

    lid = lid_loss(image_uint8, label)
    ltotal = lambda1 * l1 +  lambda2* lpips + lambda3*lid
    return l1, lpips, lid, ltotal`

**Comment by maocaixia** [comment]
Hi！When to  release code. Thank you~ @41xu
